### DISCUSSION TITLE
Model Weight Refit in FX

### DISCUSSION THREAD

**Discussion by gs-olive** [pytorch/TensorRT#1679] [RFCs]
# Model Weight Refit

## TL;DR
TensorRT supports updating engine weights _after_ compilation via the `nvinfer1::IRefitter` class, referenced here [in C++](https://docs.nvidia.com/deeplearning/tensorrt/api/c_api/classnvinfer1_1_1_i_refitter.html), and here [in Python](https://docs.nvidia.com/deeplearning/tensorrt/api/python_api/infer/Core/Refitter.html). This could be a beneficial feature to bring into Torch-TensorRT, specifically the FX path, since models which are pre-compiled and saved can be easily refit to new training weights, so long as the model architecture is unchanged.

## Goals and Usecases
Model weight refit will assist greatly in reducing time spent compiling models with Torch-TensorRT, since models would only need to be compiled _once_ per architecture, and subsequent weight updates can be propagated into the compiled model post-compilation, without the overhead of recompiling. This also enables pre-compilation of a model architecture prior to training, which could allow for inference timing estimates ahead of sometimes-lengthy model training.

## Proposed APIs / UX
Users of this feature (in FX), would compile their model via the FX API, for example with: `torch_tensorrt.fx.compile(...)`, then save their model. Then, at a later time when loading the compiled model, if the weights have updated from their original values, the use could call the Model Weight Refit function to refit the stored weights.
 
### Example Workflow
See below for a sample workflow of compiling and saving a model via FX2TRT
https://github.com/pytorch/TensorRT/blob/deda87bd04e23f0b4cc7cb20bf9be6bd76fcf743/examples/fx/fx2trt_example.py#L23-L146
The additional step, as per the proposed API would be to call:
```python
weights = {"weight0": torch.Tensor(...), ...}
##### OR #####
weights = model.state_dict()

fx2trt_model.refit_weights(weights)
```
This function would parse the input weights dictionary, determine which of those to assign to which submodule from the splitter (assuming the model was not fully compiled in TRT), and assign the weights accordingly, using the TRT Python API for TRT-accelerated modules, and the Torch API for non-accelerated modules. `weights` could potentially be the output of `model.state_dict()` after training, or a different format.

### Design
A function would be added to FX2TRT, for example, as:
```python
def refit_weights(self, weights_dict: Dict[str, Tensor]):
    ...
```
As mentioned above, `refit_weights` would parse the input weights, determine which of those to assign to which submodule in the compiled model, and assign the weights accordingly, respecting the module boundaries of TRT and Torch. It would likely need a few helper functions, for example:
```python
def verify_weight_names(self, weights_dict: Dict[str, Tensor]):
    """ Returns a bool indicating whether the set of keys in the weight dictionary
         is identical to the set of weight names in the model (weight names should agree)
    """

def assign_weights(self, weights_dict: Dict[str, Tensor]):
    """ Iterates over submodules in the model and determines which weight names correspond to 
         that submodule, then retrieves the weights from the dictionary and assigns them to the submodule
         via one of the helper functions
    """

def assign_weights_torch(self, submodule, weights_subdict):
    """ Assigns selected weights to a Torch-executed submodule
    """

def assign_weights_trt(self, submodule, weights_subdict):
    """ Assigns selected weights to a TRT-executed submodule
    """
```

### Extensions Required to Core API implementations
The existing library should not require many changes, as this add-on would simply add functionality while preserving existing core APIs.

### Details specific for TorchScript Support 
TorchScript Python API support is more challenging in this particular case, since the weight Tensor objects would need to be transferred from Python to C++. A similar design would function well, for example `ts_model.refit_weights(weights)`, however the FX path would make a better MVP since the implementations could stay strictly in Python, via the tensorrt Python API.

### Details specific for FX support 
See above

### Prototype - Small/Medium

- Determine standard paradigm for providing new weights to existing model
  - Ensure that weight names do not change between different instances of the same model
  - Check if `model.state_dict()` for a newly-trained model contains sufficient information to update the weights of an existing model
- Implement helper functions described in Design section above
- Develop method to apply weight updates to a Torch/FX-only model (no TRT)

### MVP `1.4.0` - Medium
- Using the prototype, implement full support for the `refit_weights` function, including refitting weights with multiple TRT-accelerated submodules and multiple Torch/FX non-accelerated submodules

### Extension Phase 1 [Potential] - Medium

- Scope/Explore/Implement `refit_weights(...)` for TorchScript API
  - PyBind, the library used to send data between Python and C++ does have support for dictionaries, and thus the pipeline to pass a dictionary of weights, or a `state_dict` would be feasible

**Comment by narendasan** [comment]
## Phase 1. 

Use convert_module_to_engine to produce standalone TRT engines from PyTorch Modules, and then out of library, figure out how to populate programmatically refit settings and refit the engine. 

- https://github.com/pytorch/TensorRT/blob/c7f20f87c20e307b21fe2abdc542d873240ca871/py/torch_tensorrt/dynamo/_compiler.py#L438

## Phase 2. [MVP]

Now support the same workflow but inside of the Torch-TRT context, and start to prototype the UX for end users (only needs to support fully supported models, but should work end to end) 

## Phase 3. 

Support advanced features of Torch-TensorRT, e.g. dynamic shape, fallback to pytorch, use library libraries like huggingface and torch.compile.

## Extensions

Demos of interesting workflows (LoRAs, Offsite compilation, caching etc.)
